scripts/proto/resnet50.py

```diff
@@ -1,69 +1,139 @@
-    scale_factor=1e-4, weights=weights)])
-top1_metric = lp.Metric('categorical accuracy', top1, '%')
-top5_metric = lp.Metric('top-5 categorical accuracy', top5, '%')
-
-lp.save_model('resnet50.prototext', input, dl, 256, 90, obj,
-              metrics=[top1_metric, top5_metric],
-              callbacks=[lp.CallbackPrint(), lp.CallbackTimer(),
-                         lp.CallbackDropFixedLearningRate(
-                             drop_epoch=[30, 60, 80], amt=0.1)])
+            residual = self.downsample(x)
+        return lp.Relu(lp.Add([conv, residual]))
+
+class ResBlock:
+    """ResNet block, constructed of some number of bottleneck layers.
+
+    Here we use "layer" in the PyTorch/TensorFlow/Keras sense of the
+    word, not the way LBANN uses it.
+
+    """
+
+    def __init__(self, num_layers, mid_channels, out_channels,
+                 stride, dilation=1, bn_stats_aggregation='local'):
+        self.layers = []
+        self.layers.append(ResBottleneck(
+            mid_channels, out_channels,
+            stride, dilation=dilation, downsample=True,
+            bn_stats_aggregation=bn_stats_aggregation))
+        for i in range(num_layers - 1):
+            self.layers.append(ResBottleneck(
+                mid_channels, out_channels,
+                stride=1, dilation=dilation, downsample=False,
+                bn_stats_aggregation=bn_stats_aggregation))
+
+    def __call__(self, x):
+        for l in self.layers:
+            x = l(x)
+        return x
+
+# Construct layer graph.
+input = lp.Input(io_buffer='partitioned')
+images = lp.Identity(input)
+labels = lp.Identity(input)
+conv1 = ConvBNRelu2d(64, 7, stride=2, padding=3,
+                     bn_stats_aggregation=bn_stats_aggregation)(images)
+pool1 = lp.Pooling(conv1, num_dims=2, has_vectors=False,
+                   pool_dims_i=3, pool_pads_i=1, pool_strides_i=2,
+                   pool_mode='max')
+block1 = ResBlock(blocks[0], 64, 256, stride=1,
+                  bn_stats_aggregation=bn_stats_aggregation)(pool1)
+block2 = ResBlock(blocks[1], 128, 512, stride=2,
+                  bn_stats_aggregation=bn_stats_aggregation)(block1)
+block3 = ResBlock(blocks[2], 256, 1024, stride=2,
+                  bn_stats_aggregation=bn_stats_aggregation)(block2)
+block4 = ResBlock(blocks[3], 512, 2048, stride=2,
+                  bn_stats_aggregation=bn_stats_aggregation)(block3)
+avgpool = lp.Pooling(block4, num_dims=2, has_vectors=False,
+                     pool_dims_i=7, pool_pads_i=0, pool_strides_i=1,
+                     pool_mode='average')
+fc = lp.FullyConnected(avgpool,
+                       weights=lp.Weights(initializer=lp.HeNormalInitializer()),
+                       hint_layer=labels, has_bias=False)
+softmax = lp.Softmax(fc)
+ce = lp.CrossEntropy([softmax, labels])
+top1 = lp.CategoricalAccuracy([softmax, labels])
+top5 = lp.TopKCategoricalAccuracy([softmax, labels], k=5)
+layers = list(lp.traverse_layer_graph(input))
```

Comment:
`lp.save_model` calls `traverse_layer_graph` internally-- we should do so in one place, not both.

Reply:
We need the list of layers here so that we can extract their weights and feed them into L2 regularization. I'd like to call `traverse_layer_graph` inside of `save_model` for safety. Maybe I'm being too cautious, but these graphs should not be egregiously large. If they were really so large that graph traversal were painful, imagine how long it would take for a single forward or backward prop iteration.

Reply:
Fair enough. 👍